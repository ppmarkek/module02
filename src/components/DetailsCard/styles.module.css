.detailsCard {
  position: absolute;
  right: 0;
  top: 0;
  width: 30%;
  height: 100%;
  background: #f8f9fc;
  display: flex;
  justify-content: space-around;
  z-index: 1000;
}

.closeDetails {
  position: absolute;
  top: 0;
  left: 0;
  width: 70%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  cursor: pointer;
  z-index: 999;
}

.closeDetailsButton {
  margin-top: 5px;
  height: 42px;
  width: 42px;
  cursor: pointer;
  border-radius: 10px;
  border: 1px solid #e60606;
  background: #fff;
  color: #000;
  transition: all 0.5s;
}

.closeDetailsButton:hover {
  background: #e60606;
  color: #fff;
}
