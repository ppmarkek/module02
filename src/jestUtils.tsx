import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { reducer } from './redux/index';
import { createStore } from '@reduxjs/toolkit';

export const renderWithRedux = (ui: React.JSX.Element, state = {}) => {
  const Wrapper = ({
    children,
  }: {
    children: React.ReactNode | React.ReactNode[];
  }) => (
    <Router>
      <Provider store={createStore(reducer, state)}>{children}</Provider>
    </Router>
  );

  return render(ui, { wrapper: Wrapper });
};
