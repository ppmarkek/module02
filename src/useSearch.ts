import { useSelector } from 'react-redux';
import { AppState } from './types';

export const useSearch = () => {
  const requestResults = useSelector(
    (state: AppState) => state.componentsReducer.results
  );
  const isLoading = useSelector(
    (state: AppState) => state.componentsReducer.isLoading
  );
  const details = useSelector(
    (state: AppState) => state.componentsReducer.details
  );
  const searchResult = useSelector(
    (state: AppState) => state.componentsReducer.searchResult
  );

  const isDetailsLoading = useSelector(
    (state: AppState) => state.componentsReducer.isDetailsLoading
  );

  const isShowDetails = useSelector(
    (state: AppState) => state.componentsReducer.isShowDetails
  );

  const isError = useSelector(
    (state: AppState) => state.componentsReducer.isError
  );

  return {
    requestResults,
    isLoading,
    isDetailsLoading,
    details,
    searchResult,
    isShowDetails,
    isError,
  };
};
