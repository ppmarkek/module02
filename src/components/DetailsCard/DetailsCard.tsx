import { useDispatch } from 'react-redux';
import styles from './styles.module.css';
import { setIsShowDetails } from '../../redux/componentsReducer';
import { useSearch } from '../../useSearch';
import { useRouter } from 'next/router';

export const DetailsCard = () => {
  const dispatch = useDispatch();
  const router = useRouter();
  const { details, isDetailsLoading, isShowDetails } = useSearch();

  const closeDetails = () => {
    router.push('/');
    dispatch(setIsShowDetails(false));
  };

  return (
    isShowDetails && (
      <div>
        <div className={styles.detailsCard} data-testid="details-card">
          {isDetailsLoading ? (
            <h3 data-testid="loading-details">Loading...</h3>
          ) : (
            <div>
              {details?.name && (
                <h3 data-testid="name">Name: {details?.name}</h3>
              )}
              {details?.createdAt && <p>Created: {details?.createdAt}</p>}
              {details?.updatedAt && <p>Edited: {details?.updatedAt}</p>}
              {details?.height && <p>Height: {details?.height}</p>}
              {details?.mass && <p>Mass: {details?.mass}</p>}
            </div>
          )}
          <button
            className={styles.closeDetailsButton}
            data-testid="close-details"
            onClick={() => closeDetails()}
          >
            X
          </button>
        </div>

        <div
          data-testid="details-shadow-background"
          className={styles.closeDetails}
          onClick={() => closeDetails()}
        />
      </div>
    )
  );
};
