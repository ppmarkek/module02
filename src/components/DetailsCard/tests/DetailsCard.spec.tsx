import { renderWithRedux } from '../../../jestUtils';
import { appStoreMock, useSearchHookMock } from '../../../fixtures/results';
import * as useSearch from '../../../useSearch';
import '@testing-library/jest-dom';
import { DetailsCard } from '../DetailsCard';
import { fireEvent } from '@testing-library/react';

const useRouter = jest.fn();
jest.mock('next/router', () => ({
  useRouter: () => ({
    push: useRouter,
    route: '/',
    query: {},
    asPath: '',
  }),
}));

describe('DetailsCard component', () => {
  beforeEach(() => {
    useRouter.mockReset();
  });

  it('should render correctly', () => {
    jest.spyOn(useSearch, 'useSearch').mockImplementation(() => ({
      ...useSearchHookMock,
      isShowDetails: true,
    }));

    const component = renderWithRedux(<DetailsCard />, {
      initialState: appStoreMock,
    });

    expect(component).toMatchSnapshot();
  });

  it('should navigate to the search after clicking close button', () => {
    jest.spyOn(useSearch, 'useSearch').mockImplementationOnce(() => ({
      ...useSearchHookMock,
      isShowDetails: true,
    }));

    const { getByTestId } = renderWithRedux(<DetailsCard />, {
      initialState: appStoreMock,
    });
    fireEvent.click(getByTestId('close-details'));

    expect(useRouter).toHaveBeenCalledWith('/');
  });

  it('should navigate to the search after clicking on the background', () => {
    jest.spyOn(useSearch, 'useSearch').mockImplementationOnce(() => ({
      ...useSearchHookMock,
      isShowDetails: true,
    }));

    const { getByTestId } = renderWithRedux(<DetailsCard />, {
      initialState: appStoreMock,
    });
    fireEvent.click(getByTestId('details-shadow-background'));

    expect(useRouter).toHaveBeenCalledWith('/');
  });
});
