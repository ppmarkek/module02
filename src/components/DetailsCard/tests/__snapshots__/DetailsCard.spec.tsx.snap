// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsCard component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="details-card"
          data-testid="details-card"
        >
          <div>
            <h3
              data-testid="name"
            >
              Name: 
              test name
            </h3>
            <p>
              Created: 
              01-01-0101
            </p>
            <p>
              Edited: 
              02-02-0202
            </p>
            <p>
              Height: 
              100
            </p>
            <p>
              Mass: 
              100
            </p>
          </div>
          <button
            class="close-details-button"
            data-testid="close-details"
          >
            X
          </button>
        </div>
        <div
          class="close-details"
          data-testid="details-shadow-background"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="details-card"
        data-testid="details-card"
      >
        <div>
          <h3
            data-testid="name"
          >
            Name: 
            test name
          </h3>
          <p>
            Created: 
            01-01-0101
          </p>
          <p>
            Edited: 
            02-02-0202
          </p>
          <p>
            Height: 
            100
          </p>
          <p>
            Mass: 
            100
          </p>
        </div>
        <button
          class="close-details-button"
          data-testid="close-details"
        >
          X
        </button>
      </div>
      <div
        class="close-details"
        data-testid="details-shadow-background"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
